src/ngx_http_lua_balancer.c

```diff
@@ -16,46 +16,103 @@
 #include "ngx_http_lua_directive.h"
 
 
+typedef struct {
+    ngx_uint_t                               size;
+    ngx_uint_t                               connections;
+
+    uint32_t                                 crc32;
+
+    lua_State                               *lua_vm;
+
+    ngx_queue_t                              cache;
+    ngx_queue_t                              free;
+} ngx_http_lua_balancer_keepalive_pool_t;
+
+
+typedef struct {
+    ngx_queue_t                              queue;
+    ngx_connection_t                        *connection;
+
+    ngx_http_lua_balancer_keepalive_pool_t  *cpool;
+} ngx_http_lua_balancer_keepalive_item_t;
+
+
 struct ngx_http_lua_balancer_peer_data_s {
-    /* the round robin data must be first */
-    ngx_http_upstream_rr_peer_data_t    rrp;
+    ngx_uint_t                              cpool_size;
+    ngx_uint_t                              keepalive_requests;
+    ngx_msec_t                              keepalive_timeout;
+
+    ngx_uint_t                              more_tries;
+    ngx_uint_t                              total_tries;
 
-    ngx_http_lua_srv_conf_t            *conf;
-    ngx_http_request_t                 *request;
+    int                                     last_peer_state;
 
-    ngx_uint_t                          more_tries;
-    ngx_uint_t                          total_tries;
+    uint32_t                                cpool_crc32;
 
-    struct sockaddr                    *sockaddr;
-    socklen_t                           socklen;
+    void                                   *data;
 
-    ngx_str_t                          *host;
-    in_port_t                           port;
+    ngx_event_get_peer_pt                   original_get_peer;
+    ngx_event_free_peer_pt                  original_free_peer;
+
+#if (NGX_HTTP_SSL)
+    ngx_event_set_peer_session_pt           original_set_session;
+    ngx_event_save_peer_session_pt          original_save_session;
+#endif
 
-    int                                 last_peer_state;
+    ngx_http_request_t                     *request;
+    ngx_http_lua_srv_conf_t                *conf;
+    ngx_http_lua_balancer_keepalive_pool_t *cpool;
+
+    ngx_str_t                              *host;
+
+    struct sockaddr                        *sockaddr;
+    socklen_t                               socklen;
+
+    unsigned                                keepalive:1;
 
 #if !(HAVE_NGX_UPSTREAM_TIMEOUT_FIELDS)
-    unsigned                            cloned_upstream_conf;  /* :1 */
+    unsigned                                cloned_upstream_conf;  /* :1 */
```

Comment:
We should use bit fields here since they are after the bit field above (`keepalive`).